routing/index_graph_starter_joints.hpp

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include "routing/segment.hpp"
+#include "routing/index_graph_starter.hpp"
+
+#include "geometry/point2d.hpp"
+
+#include <algorithm>
+#include <map>
+#include <queue>
+#include <set>
+#include <utility>
+
+namespace routing
+{
+class IndexGraphStarterJoints
+{
+public:
+  using Vertex = JointSegment;
+  using Edge = JointEdge;
+  using Weight = RouteWeight;
+
+  explicit IndexGraphStarterJoints(IndexGraphStarter & starter) : m_starter(starter) {}
+  IndexGraphStarterJoints(IndexGraphStarter & starter,
+                          Segment const & startSegment,
+                          Segment const & endSegment);
+
+  void Init(Segment const & startSegment, Segment const & endSegment);
+
+  JointSegment const & GetStartJoint() { return m_startJoint; }
+  JointSegment const & GetFinishJoint() { return m_endJoint; }
+
+  // These functions are A* interface.
+  RouteWeight HeuristicCostEstimate(JointSegment const & from, JointSegment const & to);
+
+  m2::PointD const & GetPoint(JointSegment const & jointSegment, bool start);
+
+  void GetOutgoingEdgesList(JointSegment const & vertex, std::vector<JointEdge> & edges)
+  {
+    GetEdgeList(vertex, true /* isOutgoing */, edges);
+  }
+
+  void GetIngoingEdgesList(JointSegment const & vertex, std::vector<JointEdge> & edges)
+  {
+    GetEdgeList(vertex, false /* isOutgoing */, edges);
+  }
+
+  WorldGraph::Mode GetMode() const { return m_starter.GetMode(); }
+  // End of A* interface.
+
+  IndexGraphStarter & GetStarter() { return m_starter; }
+
+  /// \brief Эта функция реконструирует joint по сегментам в процессе
+  ///        после построения маршрута.
+  std::vector<Segment> ReconstructJoint(JointSegment const & joint);
+
+private:
+  static auto constexpr kInvisibleId = std::numeric_limits<uint32_t>::max() - 2;
+
+  struct FakeJointSegment
+  {
+    FakeJointSegment() = default;
+    FakeJointSegment(Segment const & start, Segment const & end)
+      : m_start(start), m_end(end) {}
+
+    Segment GetSegment(bool start) const
+    {
+      return start ? m_start : m_end;
+    }
+
+    Segment m_start;
+    Segment m_end;
+  };
+
+  void AddFakeJoints(Segment const & segment, bool isOutgoing, std::vector<JointEdge> & edges);
+
+  void GetEdgeList(JointSegment const & vertex, bool isOutgoing, std::vector<JointEdge> & edges);
+
+  JointSegment CreateFakeJoint(Segment const & from, Segment const & to);
+
+  bool IsJoint(Segment const & segment, bool fromStart)
+  {
+    return m_starter.GetGraph().GetIndexGraph(segment.GetMwmId())
+                    .IsJoint(segment.GetRoadPoint(fromStart));
+  }
+
+  // Функция выполняет BFS от |startSegment| в направлении |fromStart| и ищет ближайшие joint'ы.
+  // Таким образом составляется фейковый граф joint'ов.
+  std::vector<JointEdge> FindFirstJoints(Segment const & startSegment, bool fromStart);
+
+  JointSegment CreateInvisibleJoint(Segment const & segment, bool start);
+
+  bool IsInvisible(JointSegment const & jointSegment) const
```

Comment:
Pls move methods which is not one line to cpp.

Reply:
done